File: app/models/outlet.py
from sqlalchemy import Boolean, Float, String, Text
from sqlalchemy.orm import Mapped, mapped_column, relationship

from app.database import Base


class Outlet(Base):
    __tablename__ = "outlets"

    id: Mapped[int] = mapped_column(primary_key=True)
    name: Mapped[str] = mapped_column(String(255), nullable=False)
    domain: Mapped[str] = mapped_column(String(255), unique=True, nullable=False)
    country: Mapped[str | None] = mapped_column(String(2))  # ISO 3166-1 alpha-2
    language_primary: Mapped[str | None] = mapped_column(String(10))  # BCP 47
    political_leaning_lr: Mapped[float | None] = mapped_column(Float)  # -1.0 (left) to 1.0 (right)
    political_leaning_source: Mapped[str | None] = mapped_column(String(100))
    parent_org_name: Mapped[str | None] = mapped_column(String(255))
    rss_feed_url: Mapped[str | None] = mapped_column(Text)
    active: Mapped[bool] = mapped_column(Boolean, default=True, nullable=False)

    articles: Mapped[list["Article"]] = relationship(back_populates="outlet")  # noqa: F821

File: client/lib/core/models/outlet.dart
import 'package:equatable/equatable.dart';

class OutletSummary extends Equatable {
  const OutletSummary({
    required this.id,
    required this.name,
    required this.domain,
    this.country,
    this.politicalLeaningLr,
  });

  final int id;
  final String name;
  final String domain;
  final String? country;

  // -1.0 (far left) to 1.0 (far right); null if unknown.
  final double? politicalLeaningLr;

  factory OutletSummary.fromJson(Map<String, dynamic> json) => OutletSummary(
        id: json['id'] as int,
        name: json['name'] as String,
        domain: json['domain'] as String,
        country: json['country'] as String?,
        politicalLeaningLr: (json['political_leaning_lr'] as num?)?.toDouble(),
      );

  @override
  List<Object?> get props => [id];
}

class OutletDetail extends OutletSummary {
  const OutletDetail({
    required super.id,
    required super.name,
    required super.domain,
    super.country,
    super.politicalLeaningLr,
    this.languagePrimary,
    this.politicalLeaningSource,
    this.parentOrgName,
    required this.active,
  });

  final String? languagePrimary;
  final String? politicalLeaningSource;
  final String? parentOrgName;
  final bool active;

  factory OutletDetail.fromJson(Map<String, dynamic> json) => OutletDetail(
        id: json['id'] as int,
        name: json['name'] as String,
        domain: json['domain'] as String,
        country: json['country'] as String?,
        politicalLeaningLr: (json['political_leaning_lr'] as num?)?.toDouble(),
        languagePrimary: json['language_primary'] as String?,
        politicalLeaningSource: json['political_leaning_source'] as String?,
        parentOrgName: json['parent_org_name'] as String?,
        active: json['active'] as bool,
      );

  @override
  List<Object?> get props => [id, active];
}
